presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1403,22 +1406,24 @@ private void rollbackShared()
             switch (declaredIntentionToWrite.getMode()) {
                 case STAGE_AND_MOVE_TO_TARGET_DIRECTORY:
                 case DIRECT_TO_TARGET_NEW_DIRECTORY: {
-                    // Note: there is no need to cleanup the target directory as it will only be written
-                    // to during the commit call and the commit call cleans up after failures.
-                    Path rootPath = declaredIntentionToWrite.getRootPath();
-
-                    // In the case of DIRECT_TO_TARGET_NEW_DIRECTORY, if the directory is not guaranteed to be unique
-                    // for the query, it is possible that another query or compute engine may see the directory, wrote
-                    // data to it, and exported it through metastore. Therefore it may be argued that cleanup of staging
-                    // directories must be carried out conservatively. To be safe, we only delete files that start with
-                    // the unique prefix for queries in this transaction.
-
-                    recursiveDeleteFilesAndLog(
-                            declaredIntentionToWrite.getContext(),
-                            rootPath,
-                            ImmutableList.of(declaredIntentionToWrite.getFilePrefix()),
-                            true,
-                            format("staging/target_new directory rollback for table %s", declaredIntentionToWrite.getSchemaTableName()));
+                    if (!skipTargetCleanupOnRollback || declaredIntentionToWrite.getMode() != DIRECT_TO_TARGET_NEW_DIRECTORY) {
+                        // Note: For STAGE_AND_MOVE_TO_TARGET_DIRECTORY there is no need to cleanup the target directory as it will only be written
```

Comment:
I don't quite understand this comment update? Looks like we will still cleanup the target directory for `STAGE_AND_MOVE_TO_TARGET_DIRECTORY` ?

Reply:
because for direct_to_target_new_directory, the directory you are cleaning up _is_ the target directory.   For STAGE_AND_MOVE_TO_TARGET_DIRECTORY, the directory you're cleaning up here is the staging directory, and the target directory gets cleaned up elsewhere.

Reply:
yup that's correct